### ISSUE TITLE
.broadcast get's blocked by time.sleep if it's on a MainThread

### ISSUE THREAD

**Issue by fillwithjoy1** [aaugustin/websockets#1069]
**TL:DR**: Don't use time.sleep in asyncio, use asyncio.sleep instead due to reasons listed below.

Consider this piece of code
```
websocket.broadcast(ppl, "test")
```

While this successfully broadcasts data to all clients, consider adding a time.sleep to the code would be good because logically Python executes line-by-line as this snippet below demonstrates

```
websocket.broadcast(ppl, "test") # let's broadcast important data

some_tasks() # need to complete some tasks

even_more() # complete more

even_even_more() # complete even more

time.sleep(5) # alright time to sleep for some reason
```

you can add filler code into there and still broadcast is stalled for 5 seconds, for no reason. And no the 

for e in Users:
await e.send("message")
doesn't fix the problem either

**How to avoid**

Use asyncio.sleep whenever possible as a substitute of time.sleep

~~Use threading library and whenever using both, ensure that they are not using the MainThread, otherwise this problem will indeed plague you.~~ (However as @aaugustin points out, you should avoid mixing threads unless you are skilled in using asyncio. Just learning asyncio from this library alone isn't proper enough. You can see the potential of using asyncio [here ](https://github.com/fillwithjoy1/async_io_for_beginners)in a beginner tutorial)

**Possible cause**

If you are new to asyncio, look [here](https://github.com/fillwithjoy1/async_io_for_beginners) to learn the basics. However the proper explanation is that using time.sleep in asyncio is a dumb idea, as I've learned the hard way.

~~Given how Python's async schedules task from the main event loop and async event loop generally being executed later, this would lead to time.sleep being executed first even though structurally the .broadcast would be first. (I actually don't know much how to explain, still learning asyncio). The documentation does mention that .broadcast is basically the better version of `for i in Users => Users.send()` whilst avoiding backpressure - https://websockets.readthedocs.io/en/stable/topics/broadcast.html?highlight=broadcast#backpressure-meets-broadcast.~~

oh hey issue no. 69!

**Comment by fillwithjoy1** [comment]
It's also worth noting that this issue doesn't arise when utilising threads (or not using the MainThread).

**Comment by aaugustin** [comment]
Putting a `time.sleep` anywhere in an `asyncio` application is definitely not a good idea.

There's a fairly complete discussion of this issue here: https://websockets.readthedocs.io/en/stable/topics/broadcast.html

Maybe the point you're making isn't sufficiently prominent -- I'll see if there's something to clarify.

**Comment by fillwithjoy1** [comment]
Makes sense thank you! Sent from Mail for Windows From: Aymeric AugustinSent: Thursday, 14 October 2021 5:37 PMTo: aaugustin/websocketsCc: Joy; AuthorSubject: Re: [aaugustin/websockets] .broadcast get's blocked by time.sleep if it's on a MainThread (#1069) Putting a time.sleep anywhere in an asyncio application is definitely not a good idea.There's a fairly complete discussion of this issue here: https://websockets.readthedocs.io/en/stable/topics/broadcast.htmlMaybe the point you're making isn't sufficiently prominent -- I'll see if there's something to clarify.—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by aaugustin** [comment]
https://websockets.readthedocs.io/en/stable/reference/server.html#websockets.server.WebSocketServerProtocol.close

**Comment by aaugustin** [comment]
You should avoid mixing threads and asyncio.

If you had to use multiple threads, `call_soon_threadsafe` would help, but I'm almost sure you don't.

If you need to run blocking code, look at `run_in_executor`.

**Comment by aaugustin** [comment]
I looked at the FAQ and I think that the key points are adressed so I will close this issue.
